File: tests/clients/dotnet-client-tests/ClientTests.cs
using System.Runtime.InteropServices;
using System.Text.Json;
using Xunit;

namespace Hsds.Api.Tests
{
    public class ClientTests
    {
        private static void SwitchEndianness<T>(Span<T> dataset) where T : unmanaged
        {
            var size = Marshal.SizeOf<T>();
            var dataset_bytes = MemoryMarshal.Cast<T, byte>(dataset);

            for (int i = 0; i < dataset_bytes.Length; i += size)
            {
                for (int j = 0; j < size / 2; j++)
                {
                    var i1 = i + j;
                    var i2 = i - j + size - 1;

                    byte tmp = dataset_bytes[i1];
                    dataset_bytes[i1] = dataset_bytes[i2];
                    dataset_bytes[i2] = tmp;
                }
            }
        }

        [Fact]
        public void Test()
        {
            // Arrange
            var url = new Uri("http://hsdshdflab.hdfgroup.org");
            using var client = new HsdsClient(url);
            var domainName = "/shared/tall.h5";

            // Act

            /* get domain */
            var domain = client.V2_0.Domain.GetDomain(domainName);
            var rootGroupLinks = client.V2_0.Link.GetLinks(domain.Root, domainName);

            /* get group "g1" */
            var g1_Link = rootGroupLinks.Links[0];
            var g1_Name = g1_Link.Title;

            /* get group "g1.1" */
            var g1_Links = client.V2_0.Link.GetLinks(g1_Link.Id, domainName);
            var g1_1_Link = g1_Links.Links[0];
            var g1_1_Name = g1_1_Link.Title;
            var g1_1_Links = client.V2_0.Link.GetLinks(g1_1_Link.Id, domainName);

            /* get dataset "dset1.1.1" */
            var dset_1_1_1_Link = g1_1_Links.Links[0];
            var dset_1_1_1_Name = dset_1_1_1_Link.Title;

            var dset_1_1_1_Type = client.V2_0.Dataset
                .GetDataset(dset_1_1_1_Link.Id, domainName)
                .Type;

            /* get data of dataset "dset1.1.1" as JSON */
            var jsonResponse = client.V2_0.Dataset.GetValuesAsJson(dset_1_1_1_Link.Id, domainName);

            /* get data of dataset "dset1.1.1" as stream */
            var streamResponse = client.V2_0.Dataset.GetValuesAsStream(dset_1_1_1_Link.Id, domainName);
            var stream = streamResponse.Content.ReadAsStream();

            // Assert
            Assert.Equal("g-d38053ea-3418fe27-5b08-db62bc-9076af", domain.Root);
            Assert.Equal("g1", g1_Name);
            Assert.Equal("g1.1", g1_1_Name);
            Assert.Equal("dset1.1.1", dset_1_1_1_Name);
            Assert.Equal("H5T_INTEGER", dset_1_1_1_Type.Class);
            Assert.Equal("H5T_STD_I32BE", dset_1_1_1_Type.Base);

            var expectedData = new int[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };

            /* JSON response */
            var expectedJsonDataString = JsonSerializer.Serialize(expectedData);
            var actualJsonDataString = JsonSerializer.Serialize(jsonResponse.Value[1]);
            Assert.Equal(expectedJsonDataString, actualJsonDataString);

            /* Stream response */
            var actualdata = new int[20];
            stream.ReadExactly(MemoryMarshal.AsBytes<int>(actualdata));
            SwitchEndianness<int>(actualdata);
            Assert.True(expectedData.SequenceEqual(actualdata.Skip(10)));
        }

        [Fact]
        public async Task TestAsync()
        {
            // Arrange
            var url = new Uri("http://hsdshdflab.hdfgroup.org");
            using var client = new HsdsClient(url);
            var domainName = "/shared/tall.h5";

            // Act

            /* get domain */
            var domain = await client.V2_0.Domain.GetDomainAsync(domainName);
            var rootGroupLinks = await client.V2_0.Link.GetLinksAsync(domain.Root, domainName);

            /* get group "g1" */
            var g1_Link = rootGroupLinks.Links[0];
            var g1_Name = g1_Link.Title;

            /* get group "g1.1" */
            var g1_Links = await client.V2_0.Link.GetLinksAsync(g1_Link.Id, domainName);
            var g1_1_Link = g1_Links.Links[0];
            var g1_1_Name = g1_1_Link.Title;
            var g1_1_Links = await client.V2_0.Link.GetLinksAsync(g1_1_Link.Id, domainName);

            /* get dataset "dset1.1.1" */
            var dset_1_1_1_Link = g1_1_Links.Links[0];
            var dset_1_1_1_Name = dset_1_1_1_Link.Title;

            var dset_1_1_1_Type = (await client.V2_0.Dataset
                .GetDatasetAsync(dset_1_1_1_Link.Id, domainName))
                .Type;

            /* get data of dataset "dset1.1.1" as JSON */
            var jsonResponse = await client.V2_0.Dataset.GetValuesAsJsonAsync(dset_1_1_1_Link.Id, domainName);

            /* get data of dataset "dset1.1.1" as stream */
            var streamResponse = await client.V2_0.Dataset.GetValuesAsStreamAsync(dset_1_1_1_Link.Id, domainName);
            var stream = await streamResponse.Content.ReadAsStreamAsync();

            // Assert
            Assert.Equal("g-d38053ea-3418fe27-5b08-db62bc-9076af", domain.Root);
            Assert.Equal("g1", g1_Name);
            Assert.Equal("g1.1", g1_1_Name);
            Assert.Equal("dset1.1.1", dset_1_1_1_Name);
            Assert.Equal("H5T_INTEGER", dset_1_1_1_Type.Class);
            Assert.Equal("H5T_STD_I32BE", dset_1_1_1_Type.Base);

            var expectedData = new int[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };

            /* JSON response */
            var expectedJsonDataString = JsonSerializer.Serialize(expectedData);
            var actualJsonDataString = JsonSerializer.Serialize(jsonResponse.Value[1]);
            Assert.Equal(expectedJsonDataString, actualJsonDataString);

            /* Stream response */
            var actualdata = new int[20];
            stream.ReadExactly(MemoryMarshal.AsBytes<int>(actualdata));
            SwitchEndianness<int>(actualdata);
            Assert.True(expectedData.SequenceEqual(actualdata.Skip(10)));
        }
    }
}

File: tests/clients/python-client-tests/client-tests.py
import json
import struct

import pytest
from hsds_api import HsdsClient, HsdsAsyncClient

def sync_test():

    # arrange
    url = "http://hsdshdflab.hdfgroup.org"

    with HsdsClient.create(url) as client:

        domain_name = "/shared/tall.h5"

        # act

        # get domain
        domain = client.v2_0.domain.get_domain(domain_name)
        root_group_links = client.v2_0.link.get_links(domain.root, domain_name)

        # get group "g1"
        g1_link = root_group_links.links[0]
        g1_name = g1_link.title

        # get group "g1.1"
        g1_links = client.v2_0.link.get_links(g1_link.id, domain_name)
        g1_1_link = g1_links.links[0]
        g1_1_name = g1_1_link.title
        g1_1_links = client.v2_0.link.get_links(g1_1_link.id, domain_name)

        # get dataset "dset1.1.1"
        dset_1_1_1_link = g1_1_links.links[0]
        dset_1_1_1_name = dset_1_1_1_link.title

        dset_1_1_1_type = client.v2_0.dataset \
            .get_dataset(dset_1_1_1_link.id, domain_name) \
            .type

        # get data of dataset "dset1.1.1" as JSON
        json_response = client.v2_0.dataset.get_values_as_json(dset_1_1_1_link.id, domain_name)

        # get data of dataset "dset1.1.1" as stream
        stream_response = client.v2_0.dataset.get_values_as_stream(dset_1_1_1_link.id, domain_name)
        data = stream_response.read()

        # assert
        assert "g-d38053ea-3418fe27-5b08-db62bc-9076af" == domain.root
        assert "g1" == g1_name
        assert "g1.1" == g1_1_name
        assert "dset1.1.1" == dset_1_1_1_name
        assert "H5T_INTEGER" == dset_1_1_1_type.class_
        assert "H5T_STD_I32BE" == dset_1_1_1_type.base

        expected_data = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

        # JSON response
        expected_json_data_string = json.dumps(expected_data)
        actual_json_data_string = json.dumps(json_response.value[1])
        assert expected_json_data_string == actual_json_data_string

        # Stream response
        actual_data = list(struct.unpack(f">{int(len(data)/4)}i", data))
        assert expected_data == actual_data[10:20]

@pytest.mark.asyncio
async def async_test():
    
    # arrange
    url = "http://hsdshdflab.hdfgroup.org"

    async with HsdsAsyncClient.create(url) as client:

        domain_name = "/shared/tall.h5"

        # act

        # get domain
        domain = await client.v2_0.domain.get_domain(domain_name)
        root_group_links = await client.v2_0.link.get_links(domain.root, domain_name)

        # get group "g1"
        g1_link = root_group_links.links[0]
        g1_name = g1_link.title

        # get group "g1.1"
        g1_links = await client.v2_0.link.get_links(g1_link.id, domain_name)
        g1_1_link = g1_links.links[0]
        g1_1_name = g1_1_link.title
        g1_1_links = await client.v2_0.link.get_links(g1_1_link.id, domain_name)

        # get dataset "dset1.1.1"
        dset_1_1_1_link = g1_1_links.links[0]
        dset_1_1_1_name = dset_1_1_1_link.title

        dset_1_1_1_type = (await client.v2_0.dataset \
            .get_dataset(dset_1_1_1_link.id, domain_name)) \
            .type

        # get data of dataset "dset1.1.1" as JSON
        json_response = await client.v2_0.dataset.get_values_as_json(dset_1_1_1_link.id, domain_name)

        # get data of dataset "dset1.1.1" as stream
        stream_response = await client.v2_0.dataset.get_values_as_stream(dset_1_1_1_link.id, domain_name)
        data = await stream_response.aread()

        # assert
        assert "g-d38053ea-3418fe27-5b08-db62bc-9076af" == domain.root
        assert "g1" == g1_name
        assert "g1.1" == g1_1_name
        assert "dset1.1.1" == dset_1_1_1_name
        assert "H5T_INTEGER" == dset_1_1_1_type.class_
        assert "H5T_STD_I32BE" == dset_1_1_1_type.base

        expected_data = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

        # JSON response
        expected_json_data_string = json.dumps(expected_data)
        actual_json_data_string = json.dumps(json_response.value[1])
        assert expected_json_data_string == actual_json_data_string

        # Stream response
        actual_data = list(struct.unpack(f">{int(len(data)/4)}i", data))
        assert expected_data == actual_data[10:20]